plugins/UsersManager/TokenNotificationProvider.php

```diff
@@ -44,11 +44,13 @@ public function getTokenNotificationsForDispatch(): array
         }
 
         $db = Db::get();
-        $sql = "SELECT * FROM " . Common::prefixTable('user_token_auth')
+        $sql = "SELECT * FROM " . Common::prefixTable('user_token_auth') . " AS uta"
+            . " JOIN " . Common::prefixTable('user') . " AS u ON uta.login = u.login"
             . " WHERE (date_expired is null or date_expired > ?)"
             . " AND (date_created <= ?)"
             . " AND ts_rotation_notified is null"
-            . " AND system_token = 0";
+            . " AND system_token = 0"
+            . " AND u.email != 'anonymous@example.org' AND u.invited_by IS NULL";
```

Comment:
```suggestion
            . " AND u.login != 'anonymous'";
```

Would suggest matching on the username here.

The username is fixed in various checks across the system. Changing that value in the database "converts" the user to a regular one, changing the email however does (to my knowledge) not change the behaviour of the anonymous login in any way.

The `invited_by` check is probably not something we want to have. The value is not reset if the invitation is accepted, requiring it to not be set would deactivate the notifications for nearly every user in an instance. If you actually intended the check to better match the anonymous user and just missed the brackets (`AND (u.email ... u.invited_by)`), using the username removes the necessity for that condition.

Reply:
Yeah, I wasn't sure if the user can change somehow, so added more checks than probably necessary as you mentioned the email above as well.

Changed to just check the username.